utils.py

```diff
@@ -703,3 +706,58 @@ def get_hashable_value(value):
 class OrderedCounter(collections.Counter, collections.OrderedDict):
     """Counter that remembers the order elements are first encountered."""
     pass
+
+
+def memoize(func):
+    """Decorator which provides thread-safe, cached-access to the return values
+    of function calls.
+
+    NOTE: This function uses dicts to manage the cache. This means that all
+    values provided as arguments to func *must be hashable!*
+
+    Args:
+        func: callable.
+
+    Returns:
+        callable. The same func, but calls to it using the same arguments are
+            made exactly once.
+    """
+    key_locks = {}
+    lock_for_key_locks = threading.Lock()
+    def threadsafe_access(key):
+        """Returns a threading.Lock unique to the given key."""
```

Comment:
Should use the usual Args/Returns/Raises docstring format with typeinfo, here and below.

Reply:
Done.